### ISSUE TITLE
Problem with parseJSON function

### ISSUE THREAD

**Issue by luizfelipelaviola** [date-fns/date-fns#3707]
There is an issue with the [parseJSON function](https://date-fns.org/v3.3.1/docs/parseJSON)

## The problem

The [parseJSON function](https://date-fns.org/v3.3.1/docs/parseJSON) "for convenience and ease of use these other input types are also supported via toDate: a Date instance will be cloned; a number will be treated as a timestamp", which is awesome because sometimes I receive the Date instance instead of the ISO Date string. The function still works when passing Date instance (v2.30.0 and v3.3.1), but the type definition doesn't allow the possibility of being Date instance instead of String.

### 💻 Code demonstrating the problem

```js
// v2.30.0
import { parseJSON } from 'date-fns'

const getMyAwesomeDate = (maybeDate: string | number | Date): Date => {
  return parseJSON(maybeDate) // No TS error :)
}

const result = getMyAwesomeDate(new Date().toISOString())
```

```js
// v3.3.1
import { parseJSON } from 'date-fns'

const getMyAwesomeDate = (maybeDate: string | number | Date): Date => {
  return parseJSON(maybeDate) // Argument of type 'Date' is not assignable to parameter of type 'string'.ts(2345)
}

const result = getMyAwesomeDate(new Date().toISOString())
```

### 🙁 Actual behavior

parseJSON works with ISO string, number (timestamp) or a Date, also the documentation says that is possible and it is awesome to have this possibility, useful when working with JSON string or already parsed JSON!

On version 2.30.0 we have:
`parseJSON(argument: string | number | Date): Date`

But on version 3.3.1 was change to this:
`parseJSON(dateStr: string): Date`

### 🙂 Expected behavior

`parseJSON(argument: string | number | Date): Date`

Again, the v3.3.1 documentation says that Date and number are allowed, so why typescript allows just string on this case?

## Debug information

- date-fns version: 3.3.1
- Browser/Node.js version: 20.11.0
- Your timezone: America/Sao_Paulo
- Your current time: 15:05

**Comment by EduardoBautista** [comment]
I believe the documentation should be updated to remove that it says that `Date` is allowed. This function is supposedly for, parsing a JSON value. There is no `Date` format/type in JSON so it never made sense for this function to accept a `Date` as a valid argument.

**Comment by ansgarsteinkamp** [comment]
Same problem here. According to the documentation, 'A Date instance will be cloned' and 'Any other input type or invalid date strings will return an Invalid Date' Neither is true.

**Comment by fturmel** [comment]
I agree with @EduardoBautista and think it might've even been some mistake in v2 as well because the actual type for the argument was just a string, while the JSDocs types were `string | number | Date`.

A little detective work brought me to the TypeScript migration of that function, released in v2.22.0 https://github.com/date-fns/date-fns/pull/2299#discussion_r596505736. It was decided (correctly imo) to remove the non-string argument support, but it seems like they forgot to update the docs and the JSDocs types - which importantly in v2 were the source of truth for the generated TypeScript types consumed by the users and not just part of the documentation.

https://github.com/date-fns/date-fns/blob/b3ddacb57c26b29e884456f27b08533dbdf8ea3c/src/parseJSON/index.ts#L34

https://github.com/date-fns/date-fns/blob/b3ddacb57c26b29e884456f27b08533dbdf8ea3c/src/parseJSON/index.ts#L38

In any case, I think it's clear that in v3 the argument should only be a string (the other types never made sense to begin with), and the docs should be updated accordingly. I can work on a PR later today.

The actual v3 behavior:
- a non-string argument will throw a `TypeError` (match is not a function)
- an invalid string argument will return an Invalid Date
